### ISSUE TITLE
1章p.29  preds==labels.data の .data

### ISSUE THREAD

**Issue by calneco** [YutaroOgawa/pytorch_advanced#65]
.dataがなぜ使われているのかわかりませんでした。

調べた感じでは.dataはpytorch 0.4以降　requires_grad=FalseとなったTensorを返すとありました。pytorchでtensorを扱うときにいつ.dataだったり.detach()だったりを使用するのかわかりません。
順伝播のループのなかで自動微分に影響して欲しくないから.dataを使っておられるのでしょうか？

また、detach()との違いはなんですか？
.dataは今はあまり推奨されていないとの記事も見たことがあります。

**Comment by YutaroOgawa** [comment]
@calneco さま

ありがとうございます。
おっしゃる通りで、PyTorchのバージョン0.4以降、
テンソルの値バックプロパゲーションから切り離して扱う際には、
.dataの代わりに.detach()を使用することが推奨されます。
そして、.dataは基本的に使用しません。

なぜ.dataが推奨されないかというと、.dataを使って定義した変数を変更した際に、
勾配値は更新されず、変な残り方をする現象が発生するからです。

こちらが詳しいです。
https://github.com/pytorch/pytorch/issues/6990

そのため、安全なdetach()を使うことが推奨されます。

では、本書でなぜこの部分を、labels.dataにしていて、labels.detach()にしていないかですが、
PyTorchの公式チュートリアルに合わせているからです。

https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html

公式チュートリアルでは、.dataを使用しております。
`running_corrects += torch.sum(preds == labels.data)`

第1章を読む際には、まだPyTorchに慣れていない読者も多く、
PyTorchのチュートリアルと合わせておく方が混乱が少ないと判断して、.dataにしています。

ただ、 @calneco さまのおっしゃる通り、.dataは推奨されていないので、
その点を把握されている方はぜひ、detach()の使用を、私も推奨します。

どうぞ宜しくお願い致します。
